pkg/api/endpoints/util.go

```diff
@@ -83,21 +65,63 @@ func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
 	// Next, build the N-to-M association the API wants.
 	final := []api.EndpointSubset{}
 	for key, ports := range addrSetsToPorts {
-		addrs := []api.EndpointAddress{}
-		for k := range addrSets[key] {
-			addrs = append(addrs, *k)
+		var addrs, notReadyAddrs []api.EndpointAddress
+		for k, ready := range addrSets[key] {
+			if ready {
+				addrs = append(addrs, *k)
+			} else {
+				notReadyAddrs = append(notReadyAddrs, *k)
+			}
 		}
-		final = append(final, api.EndpointSubset{Addresses: addrs, Ports: ports})
+		final = append(final, api.EndpointSubset{Addresses: addrs, NotReadyAddresses: notReadyAddrs, Ports: ports})
 	}
 
 	// Finally, sort it.
 	return SortSubsets(final)
 }
 
-type addressSet map[*api.EndpointAddress]struct{}
+// The sets of hosts must be de-duped, using IP+UID as the key.
+type addressKey struct {
```

Comment:
I think we should just get rid of addressKey and use the endpointAddress as the key itself.
We are inconsistent in the code here. we use addressKey in the map, but not during hash generation of the address set.
We should be consistent.